src/v/storage/ntp_config.h

```diff
@@ -15,18 +15,64 @@
 #include "model/metadata.h"
 #include "model/namespace.h"
 #include "ssx/sformat.h"
+#include "storage/file_sanitizer_types.h"
 #include "utils/tristate.h"
 
 #include <seastar/core/sstring.hh>
 
+#include <fmt/format.h>
+
 #include <optional>
+#include <ostream>
 
 namespace storage {
 using with_cache = ss::bool_class<struct log_cache_tag>;
 using topic_recovery_enabled
   = ss::bool_class<struct topic_recovery_enabled_tag>;
 
+/// This represents a set of capabilities that a topic can have.
+class ntp_caps {
+public:
+    explicit ntp_caps(
+      bool tiered_storage, bool iceberg, bool cloud_topics) noexcept
+      : _tiered_storage(tiered_storage)
+      , _iceberg(iceberg)
+      , _cloud_topics(cloud_topics) {}
+
+    /// Tiered storage can be enabled for this NTP
+    bool has_tiered_storage() const noexcept { return _tiered_storage; }
+    /// Iceberg can be enabled for this NTP
+    bool has_iceberg() const noexcept { return _iceberg; }
+    /// Cloud topics can be enabled for this NTP
+    bool has_cloud_topics() const noexcept { return _cloud_topics; }
+
+private:
+    bool _tiered_storage;
+    bool _iceberg;
+    bool _cloud_topics;
+};
+
 class ntp_config {
+    // Make caps based on ntp.
+    // For now we're disallowing Tiered Storage, Iceberg and Cloud Topics
+    // for internal topics and consumer offsets.
+    // The cloud topics are disabled if there is no specific override set
+    // during the topic creation because cloud topic can't be transformed
+    // into non-cloud topic and vice versa.
+    static ntp_caps
+    make_caps(const model::ntp& ntp, bool cloud_topics_override = false) {
```

Comment:
It seems odd to pass in an override, given my read on this class is it should be entirely focused on the NTP name. Plus, we check the override inline in `cloud_topic_enabled()` anyway?

Reply:
The idea is that every NTP has a set of inherent capabilities and I want to have them in one entity in the code.
For most topics this is based only on the name of the topic and its namespace. But the cloud topic is different. It's impossible to disable or enable the cloud topic property. You can only set this topic property when the topic is created. After that it can't be altered. So I thought that it'd be better to have this reflected here. Otherwise the NTP could be both TS and CT which is not correct.

Reply:
>You can only set this topic property when the topic is created.

So is ntp_caps supposed to be "based on the name, NTP is eligible for supporting feature X" or "NTP was configured with feature X"? My understanding was that `ntp_caps` encompassed the former. And that if we want something like the latter, it should be a function of the ntp config or topic_properties. Otherwise, it isn't quite clear what conditions belong in ntp_caps and which belong in ntp_configs or elsewhere